src/lib/factories.tsx

```diff
@@ -45,6 +52,34 @@ export function createShorthand(
   value?: ShorthandValue,
   options: CreateShorthandOptions = CREATE_SHORTHAND_DEFAULT_OPTIONS,
 ): React.ReactElement<Props> | null | undefined {
+  const valIsRenderFunction = typeof value === 'function' && !React.isValidElement(value)
+  if (valIsRenderFunction) {
+    const render: ShorthandRenderCallback = (shorthandValueOrRenderTree, renderTreeArg) => {
+      const shorthandArgType = {
+        isReactElement: React.isValidElement(shorthandValueOrRenderTree),
+        isRenderTreeFunc: typeof shorthandValueOrRenderTree === 'function',
+      }
+
+      const ShorthandElement = createShorthand(
+        Component,
+        mappedProp,
+        shorthandArgType.isRenderTreeFunc ? {} : (shorthandValueOrRenderTree as ShorthandValue),
+        options,
+      )
+
+      const renderTree =
+        (shorthandArgType.isRenderTreeFunc &&
+          (shorthandValueOrRenderTree as ShorthandRenderTreeFunc)) ||
+        renderTreeArg
```

Comment:
In my opinion is difficult to reason what will the value of renderTree be in this case. Maybe we can initialize it with renderTreeArg and then conditionaly change it to shorthandValueOrRenderTree? Not saying this is must but it would be helpful at least for me :)

Reply:
agree, now it is not-easy-to-grasp-the-intent expression, will rework, thanks!

Reply:
After refactoring is much better actually. No need for additional changes in my opinion.